src/core/socket.c

```diff
@@ -1398,37 +1385,81 @@ static int usbffs_dispatch_eps(SocketPort *p) {
         return r;
 }
 
+int socket_load_service_unit(Socket *s, int cfd, Unit **ret) {
+        /* Figure out what the unit that will be used to handle the connections on the socket looks like.
+         *
+         * If cfd < 0, then we don't have a connection yet. In case of Accept=yes sockets, use a fake
+         * instance name.
+         */
+
+        if (UNIT_ISSET(s->service)) {
+                *ret = UNIT_DEREF(s->service);
+                return 0;
+        }
+
+        if (!s->accept)
+                return -ENODATA;
+
+        /* Build the instance name and load the unit */
+        _cleanup_free_ char *prefix = NULL, *instance = NULL, *name = NULL;
+        int r;
+
+        r = unit_name_to_prefix(UNIT(s)->id, &prefix);
+        if (r < 0)
+                return r;
+
+        if (cfd >= 0) {
+                r = instance_from_socket(cfd, s->n_accepted, &instance);
+                if (r == -ENOTCONN)
+                        /* ENOTCONN is legitimate if TCP RST was received.
+                         * This connection is over, but the socket unit lives on. */
+                        return log_unit_debug_errno(UNIT(s), r,
+                                                    "Got ENOTCONN on incoming socket, assuming aborted connection attempt, ignoring.");
+                if (r < 0)
+                        return r;
+        }
+
+        /* For accepting sockets, we don't know how the instance will be called until we get a connection and
+         * can figure out what the peer name is. So let's use a "unknown" as the instance to make it clear
+         * that this is not an actual peer name. We also use "unknown" when we cannot figure out the peer. */
+        r = unit_name_build(prefix, instance ?: "unknown", ".service", &name);
+        if (r < 0)
+                return r;
+
+        return manager_load_unit(UNIT(s)->manager, name, NULL, NULL, ret);
+}
```

Comment:
hmm, I must say I don't find this pretty: this will make a service pop up suddenly that encapsulates an ugly implementation detail and whose weird brief extsiance might leak outside via dbus messages.

if we go this way, maybe we should have a flag for cases like this, so that we can mark the unit as "internal" or so. Units marked like that should be ignored by all dbus stuff, i.e. never generate UnitNew/UnitRemove/PropertiesChanged and never show up in any of the bus apis, such as ListUnits or so?

Reply:
I checked with the current version, and it seems no dbus messages (or any other kind) are generated for the fake name. The very first message already contains the appropriate name:
```
systemd[1]: listen1.socket: Incoming traffic
systemd[1]: Successfully forked off '(sd-accept)' as PID 3903.
systemd[1]: (sd-accept) succeeded.
systemd[1]: listen1@0-::1:2009-::1:58114.service: Trying to enqueue job listen1@0-::1:2009-::1:58114.service/start/replace
```
So I'd be inclined to leave it as is for now :) What do you think?